Implement logging for debug adapter clients

bbutkovic opened the pull request:
https://github.com/user-attachments/assets/2ffe4435-989c-4a36-bb0b-8e775c14d68f

bbutkovic committed:
Implement RPC logging for debug adapter clients

bbutkovic committed:
Implement server logs for debugger servers

RemcoSmitsDev commented:
I think should consider to always enable RPC messages, because for connecting to a debug adapter the initial request is crucial. Right now we cannot see them, because we did not enable RPC messages yet.

RemcoSmitsDev commented:
I also think we should consider refactoring how we pass around the **transport params** and **IO params,** maybe we should move the connect tcp/stdio methods to the transport layer. So each debug adapter does not have to care about how they should return the correct transport + log IO params.

Anthony-Eid commented:
This looks good! Maybe it would be beneficial to use create a Model<DapLogger> instead of using Mutex<Vec> to store the log messages. That why it would be more inline with how the rest of the Zed code base handles stuff. However, that's not a blocking issue and can be changed in the future.

Anthony-Eid commented on crates/debugger_tools/TODO.md:
```diff
@@ -0,0 +1 @@
+- [ ] housekeeping: ensure references to debug adapters are cleaned when they are closed
```
I think it would be better if we moved the TODO list to one of the zed_debugger sub channels to be more in line with the rest of the code base

Anthony-Eid commented on crates/dap/src/client.rs:
```diff
@@ -42,6 +44,8 @@ pub struct DebugAdapterClient {
     sequence_count: AtomicU64,
     config: DebugAdapterConfig,
+    io_handlers: Arc<Mutex<Vec<IoHandler>>>,
+    log_io_handlers: Option<Arc<Mutex<Vec<IoHandler>>>>,
```
I think the code would be simpler and more consistent with Zed's codebase if we create a `Model<io_handler>`. That way we can avoid management of the Mutex locks and let gpui handle it instead

bbutkovic commented:
Thanks for the input! I'll make the changes and let you know when it's finished. 

> I also think we should consider refactoring how we pass around the **transport params** and **IO params,** maybe we should move the connect tcp/stdio methods to the transport layer. So each debug adapter does not have to care about how they should return the correct transport + log IO params.

I think that's a good idea. Do you want me to make this change in my branch?

> I think should consider to always enable RPC messages, because for connecting to a debug adapter the initial request is crucial. Right now we cannot see them, because we did not enable RPC messages yet.

Just so I'm sure I got it right, we should get rid of the option to enable/disable RPC logs, right? Or should we enable it by default and allow people to disable it? I think if we're going to default to it being enabled we can just get rid of the toggle altogether.

RemcoSmitsDev committed:
This cleans up the way we pass through the input and output readers for logging. So not each debug adapters have to map the AdapterLogIO fields.  I also removed some specific when has logs from the client, because the client is not responsible for that.  Removed an not needed/duplicated dependency  Fix formatting & clippy

This cleans up the way we pass through the input and output readers for logging. So not each debug adapters have to map the AdapterLogIO fields.

I also removed some specific when has logs from the client, because the client is not responsible for that.

Removed an not needed/duplicated dependency

Fix formatting & clippy

RemcoSmitsDev committed:
Merge debugger branch

Co-Authored-By: Anthony Eid <hello@anthonyeid.me>

RemcoSmitsDev committed:
Implement `has_adapter_logs` for each transport impl

RemcoSmitsDev committed:
Make adapter stdout logging work

RemcoSmitsDev committed:
Add conditional render for adapter log back

RemcoSmitsDev committed:
Oops forgot to pipe the output

RemcoSmitsDev committed:
Always enable rpc messages

Previously, RPC messages were only stored when explicitly enabled, which occurred after the client was already running. This approach prevented debugging of requests sent during the initial connection period. By always enabling RPC messages, we ensure that all requests, including those during the connection phase, are captured and available for debugging purposes.

This could help use debug when someone has troble getting a debug starting. This improvement could be particularly helpful in debugging scenarios where users encounter issues during the initial connection or startup phase of their debugging sessions.

RemcoSmitsDev commented:
Thanks a lot🔥
